File: src/main/java/ecorayen/models/badge.java
package ecorayen.models;

import java.io.File;
import java.io.Serializable;
import java.util.Objects;

public class badge implements Serializable {

    private Integer id;
    private String name;
    private String description;
    private String imagePath;
    private File imageFile;
    // Default constructor
    public badge() {
    }

    // Constructor with fields (excluding id, imageFile, and challenge)
    public badge(String name, String description, String imagePath) {
        this.name = name;
        this.description = description;
        this.imagePath = imagePath;
    }

    // Constructor with all fields (excluding challenge)
    public badge(Integer id, String name, String description, String imagePath, File imageFile) {
        this.id = id;
        this.name = name;
        this.description = description;
        this.imagePath = imagePath;
        this.imageFile = imageFile;
    }

    // Getters and Setters

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getImagePath() {
        return imagePath;
    }

    public void setImagePath(String imagePath) {
        this.imagePath = imagePath;
    }

    // public challenge getChallenge() { // Removed getChallenge()
    //     return challenge;
    // }

    // public void setChallenge(challenge challenge) { // Removed setChallenge()
    //     this.challenge = challenge;
    // }

    public File getImageFile() {
        return imageFile;
    }

    public void setImageFile(File imageFile) {
        this.imageFile = imageFile;
        // If a new file is uploaded, you can handle any additional logic here
        if (imageFile != null) {

        }
    }

    @Override
    public String toString() {
        return "badge{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", description='" + description + '\'' +
                ", imagePath='" + imagePath + '\'' +
                // ", challenge=" + challenge + // Removed challenge from toString
                ", imageFile=" + imageFile +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        badge badge = (badge) o;
        return Objects.equals(id, badge.id) && Objects.equals(name, badge.name) && Objects.equals(description, badge.description) && Objects.equals(imagePath, badge.imagePath) && Objects.equals(imageFile, badge.imageFile);

    }

    @Override
    public int hashCode() {
        return Objects.hash(id, name, description, imagePath, imageFile);

    }
}

File: src/Entity/Badge.php
<?php
namespace App\Entity;

use App\Repository\BadgeRepository;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\HttpFoundation\File\File;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: BadgeRepository::class)]
class Badge
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 255)]
    private ?string $name = null;

    #[ORM\Column(length: 255)]
    private ?string $description = null;

    #[ORM\Column(length: 255, nullable: true)]
    private ?string $imagePath = null; // Property to store the path of the uploaded image

    #[ORM\ManyToOne(inversedBy: 'badge')]
    private ?Challenge $challenge = null;

    // Property to hold the uploaded file temporarily
    #[Assert\File(mimeTypes: ['image/jpeg', 'image/png', 'image/gif'])]
    private ?File $imageFile = null;

    // Getters and Setters

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getName(): ?string
    {
        return $this->name;
    }

    public function setName(string $name): static
    {
        $this->name = $name;
        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(string $description): static
    {
        $this->description = $description;
        return $this;
    }

    public function getImagePath(): ?string
    {
        return $this->imagePath;
    }

    public function setImagePath(?string $imagePath): static
    {
        $this->imagePath = $imagePath;
        return $this;
    }

    public function getImageFile(): ?File
    {
        return $this->imageFile;
    }

    public function setImageFile(?File $imageFile = null): static
    {
        $this->imageFile = $imageFile;

        // If a new file is uploaded, you can handle any additional logic here
        if ($imageFile) {
            // For example, you could set an updatedAt property if you have one
        }

        return $this;
    }

    public function getChallenge(): ?Challenge
    {
        return $this->challenge;
    }

    public function setChallenge(?Challenge $challenge): static
    {
        $this->challenge = $challenge;
        return $this;
    }
}